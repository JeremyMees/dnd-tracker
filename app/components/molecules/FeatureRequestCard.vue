<script setup lang="ts">
const emit = defineEmits<{
  update: [FeatureVotes]
  login: []
}>()

const props = defineProps<{ feature: FeatureRequest }>()

const { t } = useI18n()
const { user } = useAuthentication()

const hasVoted = computed<FeatureVote | undefined>(() => {
  if (!user.value) return undefined
  else if (props.feature.voted.like.includes(user.value.id)) return 'like'
  else if (props.feature.voted.dislike.includes(user.value.id)) return 'dislike'
  else return undefined
})

function toggleVote(vote: FeatureVote): void {
  if (!user.value) return

  const userId = user.value.id
  const votes = {
    like: [...props.feature.voted.like],
    dislike: [...props.feature.voted.dislike],
  }

  if (hasVoted.value === vote) {
    votes[vote] = votes[vote].filter(id => id !== userId)
  }
  else {
    const oppositeVote: FeatureVote = vote === 'like' ? 'dislike' : 'like'
    if (!votes[vote].includes(userId)) votes[vote].push(userId)
    votes[oppositeVote] = votes[oppositeVote].filter(id => id !== userId)
  }

  emit('update', votes)
}
</script>

<template>
  <UiCard>
    <UiCardHeader>
      <UiCardTitle>
        <div class="flex items-center gap-x-2">
          <UiAvatar class="border-2 border-background">
            <UiAvatarImage
              v-tippy="feature.created_by.username"
              :src="feature.created_by.avatar"
              :alt="feature.created_by.username"
            />
            <UiAvatarFallback>
              <Icon
                name="tabler:user"
                class="size-6 min-w-6 text-muted-foreground"
              />
            </UiAvatarFallback>
          </UiAvatar>
          <h3 data-test-title>
            {{ feature.title }}
            <NuxtTime
              class="text-xs ml-2 text-muted-foreground font-normal"
              :datetime="feature.created_at"
              month="numeric"
              day="numeric"
              year="numeric"
            />
          </h3>
        </div>
      </UiCardTitle>
    </UiCardHeader>
    <UiCardContent>
      <div class="flex gap-x-4 flex-1">
        <div class="flex flex-col gap-y-1">
          <button
            data-test-like-button
            :disabled="feature.status === 'added'"
            class="bg-background/50 border-4 border-background rounded-lg py-1 px-2 flex flex-col items-center disabled:opacity-60 disabled:cursor-not-allowed"
            :class="{
              '!bg-primary/50 !border-primary': hasVoted === 'like',
            }"
            @click="user ? toggleVote('like') : emit('login')"
          >
            <Icon
              name="tabler:thumb-up"
              aria-hidden="true"
            />
            <span
              data-test-like-count
              class="font-bold text-xs"
            >
              {{ feature.voted.like.length }}
            </span>
          </button>
          <button
            data-test-dislike-button
            :disabled="feature.status === 'added'"
            class="bg-background/50 border-4 border-background rounded-lg py-1 px-2 flex flex-col items-center disabled:opacity-60 disabled:cursor-not-allowed"
            :class="{
              '!bg-primary/50 !border-primary': hasVoted === 'dislike',
            }"
            @click="user ? toggleVote('dislike') : emit('login')"
          >
            <span
              data-test-dislike-count
              class="font-bold text-xs"
            >
              {{ feature.voted.dislike.length }}
            </span>
            <Icon
              name="tabler:thumb-down"
              aria-hidden="true"
            />
          </button>
        </div>

        <div class="flex flex-col flex-1 items-start justify-start gap-y-4">
          <div
            v-if="feature.status !== 'accepted'"
            data-test-status
            class="text-sm px-2 py-1 rounded-lg w-fit border-2"
            :class="{
              'bg-info/50 border-info': feature.status === 'review',
              'bg-secondary/50 border-secondary': feature.status === 'progress',
              'bg-success/50 border-success': feature.status === 'added',
            }"
          >
            {{ t(`pages.featureRequest.status.${feature.status}`) }}
          </div>
          <p
            data-test-text
            class="text-sm"
          >
            {{ feature.text }}
          </p>
        </div>
      </div>
    </UiCardContent>
  </UiCard>
</template>
