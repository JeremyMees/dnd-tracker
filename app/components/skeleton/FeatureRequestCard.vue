<template>
  <UiCardBase>
    <UiCardHeader>
      <UiCardTitle>
        <div class="flex items-center gap-x-2">
          <UiSkeletonBase class="size-9 rounded-full" />
          <UiSkeletonBase class="w-[200px] h-6 rounded-full" />
        </div>
      </UiCardTitle>
    </UiCardHeader>
    <UiCardContent>
      <div class="flex gap-x-4 flex-1">
        <div class="flex flex-col gap-y-1">
          <UiSkeletonBase class="w-10 h-12 rounded-LG" />
          <UiSkeletonBase class="w-10 h-12 rounded-lg" />
        </div>
        <UiSkeletonBase class="w-full rounded-lg" />
      </div>
    </UiCardContent>
  </UiCardBase>
</template>
