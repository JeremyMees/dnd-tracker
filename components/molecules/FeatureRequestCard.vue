<script setup lang="ts">
const emit = defineEmits<{
  update: [FeatureVotes]
  login: []
}>()

const props = defineProps<{ feature: FeatureRequest }>()

const { locale, t } = useI18n()
const profile = useProfile()

const date = computed<string>(() => {
  const dateString = new Date(props.feature.created_at)
  return dateString.toLocaleDateString(locale.value === 'nl' ? 'nl-NL' : 'en-US')
})

const hasVoted = computed<FeatureVote | undefined>(() => {
  if (!profile.data) return undefined
  else if (props.feature.voted.like.includes(profile.data.id)) return 'like'
  else if (props.feature.voted.dislike.includes(profile.data.id)) return 'dislike'
  else return undefined
})

function toggleVote(vote: FeatureVote): void {
  if (!profile.data) return

  if (hasVoted.value === vote) {
    props.feature.voted[vote] = props.feature.voted[vote].filter(id => id !== profile.data!.id)
  }
  else props.feature.voted[vote].push(profile.data.id)

  const reset: FeatureVote = vote === 'like' ? 'dislike' : 'like'
  props.feature.voted[reset] = props.feature.voted[reset].filter(id => id !== profile.data!.id)

  emit('update', props.feature.voted)
}
</script>

<template>
  <UiCard>
    <UiCardHeader>
      <UiCardTitle>
        <div class="flex items-center gap-x-2">
          <UiAvatar class="border-2 border-background">
            <UiAvatarImage
              v-tippy="feature.created_by.username"
              :src="feature.created_by.avatar"
              :alt="feature.created_by.username"
            />
            <UiAvatarFallback>
              <Icon
                name="tabler:user"
                class="size-6 min-w-6 text-muted-foreground"
              />
            </UiAvatarFallback>
          </UiAvatar>
          <h2>
            {{ feature.title }}
            <span class="body-extra-small ml-2 text-muted-foreground">
              {{ date }}
            </span>
          </h2>
        </div>
      </UiCardTitle>
    </UiCardHeader>
    <UiCardContent>
      <div class="flex gap-x-4 flex-1">
        <div class="flex flex-col gap-y-1">
          <button
            class="bg-background/50 border-4 border-background rounded-lg py-1 px-2 flex flex-col items-center"
            :class="{
              '!bg-primary/50 !border-primary': hasVoted === 'like',
            }"
            @click="profile.data ? toggleVote('like') : emit('login')"
          >
            <Icon
              name="tabler:thumb-up"
              aria-hidden="true"
            />
            <span class="font-bold text-xs">
              {{ feature.voted.like.length }}
            </span>
          </button>
          <button
            class="bg-background/50 border-4 border-background rounded-lg py-1 px-2 flex flex-col items-center"
            :class="{
              '!bg-primary/50 !border-primary': hasVoted === 'dislike',
            }"
            @click="profile.data ? toggleVote('dislike') : emit('login')"
          >
            <span class="font-bold text-xs">
              {{ feature.voted.dislike.length }}
            </span>
            <Icon
              name="tabler:thumb-down"
              aria-hidden="true"
            />
          </button>
        </div>

        <div class="flex flex-col flex-1 items-start justify-start gap-y-4">
          <div
            v-if="feature.status !== 'accepted'"
            class="body-small px-2 py-1 rounded-lg w-fit border-2"
            :class="{
              'bg-info/50 border-info': feature.status === 'review',
              'bg-success/50 border-success': feature.status === 'progress',
            }"
          >
            {{ t(`pages.featureRequest.status.${feature.status}`) }}
          </div>
          <p class="body-small">
            {{ feature.text }}
          </p>
        </div>
      </div>
    </UiCardContent>
  </UiCard>
</template>
