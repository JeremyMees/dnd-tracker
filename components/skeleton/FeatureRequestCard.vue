<template>
  <UiCard>
    <UiCardHeader>
      <UiCardTitle>
        <div class="flex items-center gap-x-2">
          <UiSkeleton class="size-9 rounded-full" />
          <UiSkeleton class="w-[200px] h-6 rounded-full" />
        </div>
      </UiCardTitle>
    </UiCardHeader>
    <UiCardContent>
      <div class="flex gap-x-4 flex-1">
        <div class="flex flex-col gap-y-1">
          <UiSkeleton class="w-10 h-12 rounded-LG" />
          <UiSkeleton class="w-10 h-12 rounded-lg" />
        </div>
        <UiSkeleton class="w-full rounded-lg" />
      </div>
    </UiCardContent>
  </UiCard>
</template>
