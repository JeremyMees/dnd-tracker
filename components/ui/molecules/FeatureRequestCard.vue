<script setup lang="ts">
const emit = defineEmits<{
  update: [FeatureVotes]
  login: []
}>()

const props = defineProps<{ feature: FeatureRequest }>()

const { locale, t } = useI18n()
const profile = useProfile()

const date = computed<string>(() => {
  const dateString = new Date(props.feature.created_at)
  return dateString.toLocaleDateString(locale.value === 'nl' ? 'nl-NL' : 'en-US')
})

const hasVoted = computed<FeatureVote | undefined>(() => {
  if (!profile.data) return undefined
  else if (props.feature.voted.like.includes(profile.data.id)) return 'like'
  else if (props.feature.voted.dislike.includes(profile.data.id)) return 'dislike'
  else return undefined
})

function toggleVote(vote: FeatureVote): void {
  if (!profile.data) return

  if (hasVoted.value === vote) {
    props.feature.voted[vote] = props.feature.voted[vote].filter(id => id !== profile.data!.id)
  }
  else props.feature.voted[vote].push(profile.data.id)

  const reset: FeatureVote = vote === 'like' ? 'dislike' : 'like'
  props.feature.voted[reset] = props.feature.voted[reset].filter(id => id !== profile.data!.id)

  emit('update', props.feature.voted)
}
</script>

<template>
  <Card
    color="background"
    class="flex flex-col gap-4"
  >
    <div class="flex gap-x-4 flex-1">
      <div class="flex flex-col gap-y-1">
        <button
          class="bg-black/50 border-4 border-black rounded-lg py-1 px-2 flex flex-col items-center"
          :class="{
            '!bg-primary/50 !border-primary': hasVoted === 'like',
          }"
          @click="profile.data ? toggleVote('like') : emit('login')"
        >
          <Icon
            name="tabler:thumb-up"
            aria-hidden="true"
          />
          <span class="font-bold">
            {{ feature.voted.like.length }}
          </span>
        </button>
        <button
          class="bg-black/50 border-4 border-black rounded-lg py-1 px-2 flex flex-col items-center"
          :class="{
            '!bg-primary/50 !border-primary': hasVoted === 'dislike',
          }"
          @click="profile.data ? toggleVote('dislike') : emit('login')"
        >
          <span class="font-bold">
            {{ feature.voted.dislike.length }}
          </span>
          <Icon
            name="tabler:thumb-down"
            aria-hidden="true"
          />
        </button>
      </div>

      <div class="flex flex-col flex-1 items-start justify-between gap-y-3">
        <div class="space-y-2">
          <Avatar
            :img="feature.created_by.avatar"
            :username="feature.created_by.username"
          />
          <h2 class="py-2">
            {{ feature.title }}
            <span class="body-extra-small ml-2 bg-slate-700 rounded-lg px-2 py-1">
              {{ date }}
            </span>
          </h2>
          <p
            v-if="feature.status !== 'accepted'"
            class="body-small"
            :class="{
              'text-info': feature.status === 'review',
              'text-success': feature.status === 'progress',
            }"
          >
            {{ t(`pages.featureRequest.status.${feature.status}`) }}
          </p>
          <p class="body-small">
            {{ feature.text }}
          </p>
        </div>
      </div>
    </div>
  </Card>
</template>
