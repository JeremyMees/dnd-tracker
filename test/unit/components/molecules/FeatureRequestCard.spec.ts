import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'
import FeatureRequestCard from '~/components/molecules/FeatureRequestCard.vue'
import SocialProfile from '~~/test/unit/fixtures/social-profile.json'

interface Props {
  feature: FeatureRequest
}

const props: Props = {
  feature: {
    voted: {
      like: ['id1', 'id2'],
      dislike: ['id3', 'id4'],
    },
    created_by: SocialProfile,
    status: 'review',
    text: 'text',
    created_at: '2021-01-01',
    id: 1,
    title: 'title',
  },
}

describe('FeatureRequestCard', async () => {
  vi.mock('~/composables/useAuthentication', () => ({
    useAuthentication: () => ({
      user: ref({ id: 'test-user-id' }),
    }),
  }))

  it('Should match snapshot', async () => {
    const component = await mountSuspended(FeatureRequestCard, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should render correct with default props', async () => {
    const component = await mountSuspended(FeatureRequestCard, { props })

    expect(component.find('[data-test-title]').text()).toContain(props.feature.title)
    expect(component.find('[data-test-status]').text()).toBe(`pages.featureRequest.status.${props.feature.status}`)
    expect(component.find('[data-test-like-button]').exists()).toBeTruthy()
    expect(component.find('[data-test-like-button]').attributes('class')).not.toContain('!bg-primary/50 !border-primary')
    expect(component.find('[data-test-like-count]').text()).toBe(props.feature.voted.like.length.toString())
    expect(component.find('[data-test-dislike-button]').exists()).toBeTruthy()
    expect(component.find('[data-test-dislike-button]').attributes('class')).not.toContain('!bg-background/50 !border-background')
    expect(component.find('[data-test-dislike-count]').text()).toBe(props.feature.voted.dislike.length.toString())
    expect(component.find('[data-test-text]').text()).toBe(props.feature.text)
  })

  it('Should not render label when status is accepted', async () => {
    const component = await mountSuspended(FeatureRequestCard, { props: { ...props, feature: { ...props.feature, status: 'accepted' } } })

    expect(component.find('[data-test-status]').exists()).toBeFalsy()
  })

  it('Should be able to toggle vote', async () => {
    const component = await mountSuspended(FeatureRequestCard, { props })
    const likeButton = component.find('[data-test-like-button]')

    await likeButton.trigger('click')
    await nextTick()

    expect(component.emitted('update')?.[0]).toBeTruthy()

    await component.setProps({
      feature: {
        ...props.feature,
        voted: {
          like: [...props.feature.voted.like, 'test-user-id'],
          dislike: [...props.feature.voted.dislike],
        },
      },
    })

    expect(likeButton.attributes('class')).toContain('!bg-primary/50 !border-primary')

    await likeButton.trigger('click')
    await nextTick()

    expect(component.emitted('update')?.[1]).toBeTruthy()

    await component.setProps({
      feature: {
        ...props.feature,
        voted: {
          like: props.feature.voted.like.filter(id => id !== 'test-user-id'),
          dislike: [...props.feature.voted.dislike],
        },
      },
    })

    expect(likeButton.attributes('class')).not.toContain('!bg-primary/50 !border-primary')
  })
})
