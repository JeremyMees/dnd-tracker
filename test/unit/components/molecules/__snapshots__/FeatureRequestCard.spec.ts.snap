// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeatureRequestCard > Should match snapshot 1`] = `
"<div class="rounded-lg border-4 border-secondary bg-secondary/50 text-card-foreground shadow-sm">
  <div class="flex flex-col gap-y-1.5 p-6">
    <h3 class="text-2xl font-semibold leading-none tracking-tight">
      <div class="flex items-center gap-x-2"><span class="inline-flex items-center justify-center font-normal text-foreground select-none shrink-0 bg-secondary overflow-hidden h-10 w-10 text-xs rounded-full border-2 border-background"><img role="img" src="avatar" class="h-full w-full object-cover" alt="username" style="display: none;"><span><span class="iconify i-tabler:user size-6 min-w-6 text-muted-foreground" aria-hidden="true"></span></span></span>
        <h3 data-test-title="">title <time datetime="2021-01-01T00:00:00.000Z" class="text-xs ml-2 text-muted-foreground font-normal">1/1/2021</time></h3>
      </div>
    </h3>
  </div>
  <div class="p-6 pt-0">
    <div class="flex gap-x-4 flex-1">
      <div class="flex flex-col gap-y-1"><button data-test-like-button="" class="bg-background/50 border-4 border-background rounded-lg py-1 px-2 flex flex-col items-center disabled:opacity-60 disabled:cursor-not-allowed"><span class="iconify i-tabler:thumb-up" aria-hidden="true"></span><span data-test-like-count="" class="font-bold text-xs">2</span></button><button data-test-dislike-button="" class="bg-background/50 border-4 border-background rounded-lg py-1 px-2 flex flex-col items-center disabled:opacity-60 disabled:cursor-not-allowed"><span data-test-dislike-count="" class="font-bold text-xs">2</span><span class="iconify i-tabler:thumb-down" aria-hidden="true"></span></button></div>
      <div class="flex flex-col flex-1 items-start justify-start gap-y-4">
        <div data-test-status="" class="text-sm px-2 py-1 rounded-lg w-fit border-2 bg-info/50 border-info">pages.featureRequest.status.review</div>
        <p data-test-text="" class="text-sm">text</p>
      </div>
    </div>
  </div>
</div>"
`;
